### ISSUE TITLE
OS X installer pkg's own Bom file does not list contents of Payload

### ISSUE THREAD

**Issue by jps3** [facebook/osquery#1843]
Oddball issue(?) which I have noticed for awhile, but have built up enough curiosity to ask about, since I could not find it already listed in any open/closed issues. (If I missed, my apologies.)

Download current osquery-X.Y.Z.pkg:
```
$ wget https://osquery-packages.s3.amazonaws.com/darwin/osquery-1.7.0.pkg
--2016-02-11 14:26:36--  https://osquery-packages.s3.amazonaws.com/darwin/osquery-1.7.0.pkg
Resolving osquery-packages.s3.amazonaws.com... 54.231.80.88
Connecting to osquery-packages.s3.amazonaws.com|54.231.80.88|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 6276539 (6.0M) [application/octet-stream]
Saving to: 'osquery-1.7.0.pkg'

osquery-1.7.0.pkg                        100%[================================================================================>]   5.99M  31.0MB/s    in 0.2s

2016-02-11 14:26:37 (31.0 MB/s) - 'osquery-1.7.0.pkg' saved [6276539/6276539]
```

Then expand the *.pkg to a subdirectory, and use `lsbom` command to list the contents of the pkg's Bom file:
```
$  pkgutil --expand osquery-1.7.0.pkg pkg_dir
$   ls -l pkg_dir/
total 12312
-rw-r--r--  1 user  staff    35305 Feb  1 23:36 Bom
-rw-r--r--  1 user  staff      467 Feb  1 23:36 PackageInfo
-rw-r--r--  1 user  staff  6258793 Feb  1 23:36 Payload
drwxr-xr-x  2 user  staff       68 Feb 11 14:26 Scripts
$   lsbom osquery-1.7.0.pkg.d/Bom
.       40755   501/20
./Bom   100644  501/20  512     1313622608
./PackageInfo   100644  501/20  467     462450245
./Payload       100644  501/20  6258793 3046386822
./Scripts       40755   501/20
```

This can also be seen using `pkgutil` for an already-installed pkg:
```
$ pkgutil --lsbom com.facebook.osquery
.
./Bom
./PackageInfo
./Payload
./Scripts
```

While not critical, it could be an issue with any kind of installation management (ex. Munki) which utilizes the lsbom data to determine what to remove.

Just thought I'd point it out and ask.

(Also, thanks for such a great product, I have greatly enjoyed using osquery and seeing it evolve!)

**Comment by theopolis** [comment]
Interesting, any idea why this isn't listing the contents? Here's the package build script: https://github.com/facebook/osquery/blob/master/tools/deployment/make_osx_package.sh#L216

To test you can:
```
make -j 5
make packages
```

**Comment by jps3** [comment]
I will give building a try and see if I can figure something out.

**Comment by sharvilshah** [comment]
So, I created a package (`make packages`) from my local checkout and expanded that using `pkgutil`.

The `Bom` file there lists the contents of `Payload`.

```
❯ tree /tmp/osq_pkg
/tmp/osq_pkg
├── Bom
├── PackageInfo
├── Payload
└── Scripts

1 directory, 3 files

❯ lsbom -s /tmp/osq_pkg/Bom
.
./private
./private/var
./private/var/log
./private/var/log/osquery
./private/var/osquery
./private/var/osquery/com.facebook.osqueryd.conf
./private/var/osquery/com.facebook.osqueryd.plist
./private/var/osquery/osquery.example.conf
./private/var/osquery/packs
./private/var/osquery/packs/hardware-monitoring.conf
./private/var/osquery/packs/incident-response.conf
./private/var/osquery/packs/it-compliance.conf
./private/var/osquery/packs/osquery-monitoring.conf
./private/var/osquery/packs/osx-attacks.conf
./private/var/osquery/packs/vuln-management.conf
./usr
./usr/local
./usr/local/bin
./usr/local/bin/osqueryctl
./usr/local/bin/osqueryd
./usr/local/bin/osqueryi
```

I am guessing the only difference between the package I created and the packages at osquery.io is that they are signed. Do you know if code-signed packages behave differently w.r.t `Bom` file?

**Comment by jps3** [comment]
Okay, took me awhile to get a chance to build and test this. My results are the same as sharvilshah where the expanded *.pkg file structure is correct, and the contents of the `Bom` file are correct also.

In addition I did a test having also used `productsign` to sign the resulting osquery-1.7.0-68-*.pkg. When expanded, that too had the correct contents, and it's Bom file also showed correctly.

I'm not sure what else to check, and have no idea offhand why the distributed (official) builds have incorrect Bom files.

**Comment by theopolis** [comment]
I see the problem, but I'm not yet sure how to solve, here's the script:
```bash
# 'Local' function to sign PKG packages (binaries/extensions), parameters:
# $1 - Name of keychain identity for signing binaries
# $2 - Name of keychain identity for signing packages
# $3 - Name of keychain identity for signing extensions
# $4 - /path/to/package
function sign {
  IDENTITY_BIN=$1
  IDENTITY_PKG=$2
  IDENTITY_EXT=$3
  PKG_FILE=$4

  # Extract the package data.
  log "Using identity '$IDENTITY_PKG' to sign: $PKG_FILE"
  log "Expanding package metadata and payload..."
  PKG_NAME=`basename $PKG_FILE`
  sudo rm -rf $WORKING_DIR/$PKG_NAME
  log "Expanding '$PKG_FILE' into '$WORKING_DIR/$PKG_NAME'"
  pkgutil --expand $PKG_FILE $WORKING_DIR/$PKG_NAME
  mkdir -p $WORKING_DIR/$PKG_NAME/root

  pushd $WORKING_DIR/$PKG_NAME/root
  # Extract the package payload.
  log "Extracting package payload..."
  cat ../Payload | gzip -d | cpio -id

  # Find all binary executables.
  BINARIES=`find . -type f -perm -u+x | xargs file | grep Mach-O | awk -F: '{print $1}'`
  for BIN in $BINARIES; do
    if [[ $BIN != "" ]]; then
      log "Signing binary: $BIN"
      codesign --keychain $KEYCHAIN -s "$IDENTITY_BIN" --timestamp "$BIN"
    fi
  done

  # Find all kernel extensions.
  KEXTS=`find . -type d | grep -E ".*ext$"`
  for EXT in ${KEXTS}; do
    if [[ $EXT != "" ]]; then
      log "Signing kernel extension: $EXT"
      codesign --keychain $KEYCHAIN -s "$IDENTITY_EXT" --timestamp "$EXT"
    fi
  done

  # Regenerate the package payload.
  log "Regenerating package payload..."
  sudo chown -R root:wheel .
  find . | cpio -o --format odc | gzip -c > ../Payload
  popd

  # Regenerate the package bill of materials.
  sudo rm -rf $WORKING_DIR/$PKG_NAME/root
  rm $WORKING_DIR/$PKG_NAME/Bom
  mkbom $WORKING_DIR/$PKG_NAME $WORKING_DIR/$PKG_NAME/Bom

  # Regenerate the package.
  log "Rebuilding package with signed binaries: $PKG_NAME.rebuilt"
  pkgutil --flatten $WORKING_DIR/$PKG_NAME $WORKING_DIR/$PKG_NAME.rebuilt

  # Sign the package itself
  log "Signing package: $WORKING_DIR/$PKG_NAME.signed"
  productsign --keychain $KEYCHAIN --sign "$IDENTITY_PKG" --timestamp $WORKING_DIR/$PKG_NAME.rebuilt $WORKING_DIR/$PKG_NAME.signed
  mv $WORKING_DIR/$PKG_NAME.signed $SIGNED_PACKAGES/$PKG_NAME
  rm -rf $WORKING_DIR/$PKG_NAME*
}

mkdir -p $WORKING_DIR
mkdir -p $SIGNED_PACKAGES

security unlock-keychain -p $PASSWORD $KEYCHAIN
sign "$BINARY_IDENTITY" "$PACKAGE_IDENTITY" "$EXTENSION_IDENTITY" "$1"
```

The issue is, the script **also** signs the included binary executables. It does this without trying to extract the pkg meta-information. So the rebuild uses `pkgutil --flatten ...` which is not taking a root input.

**Comment by theopolis** [comment]
If I switch around the logic here:
```bash
  # Regenerate the package bill of materials.
  sudo rm -rf $WORKING_DIR/$PKG_NAME/root
  rm $WORKING_DIR/$PKG_NAME/Bom
  mkbom $WORKING_DIR/$PKG_NAME $WORKING_DIR/$PKG_NAME/Bom
```

to:
```bash
  # Regenerate the package bill of materials.
  rm $WORKING_DIR/$PKG_NAME/Bom
  mkbom $WORKING_DIR/$PKG_NAME/root $WORKING_DIR/$PKG_NAME/Bom
  sudo rm -rf $WORKING_DIR/$PKG_NAME/root
```

We should be in business.

**Comment by jps3** [comment]
Nice find! Thank you. While I know this isn't critical, it *is* really helpful when deploying in a managed environment. The fix is much appreciated.

**Comment by sharvilshah** [comment]
This seems fixed with 1.7.1 release, thanks @theopolis!

**Comment by theopolis** [comment]
You got it! Excellent find!
